mm2src/coins/coin_errors.rs

```diff
@@ -0,0 +1,114 @@
+use crate::{utxo::rpc_clients::UtxoRpcError, DelegationError, NumConversError, UnexpectedDerivationMethod,
+            WithdrawError};
+use futures01::Future;
+use mm2_err_handle::prelude::MmError;
+use spv_validation::helpers_validation::SPVError;
+
+pub type ValidatePaymentFut<T> = Box<dyn Future<Item = T, Error = MmError<ValidatePaymentError>> + Send>;
+
+#[derive(Debug, Display, PartialEq)]
+pub enum ValidatePaymentError {
+    InternalError(String),
+    InvalidPaymentTxData(String),
+    InvalidInput(String),
+    ScriptHashTypeNotSupported(String),
+    SPVError(SPVError),
+    TransportError(String),
+    UnexpectedDerivationMethod(UnexpectedDerivationMethod),
+    UnexpectedPaymentState(String),
+    WrongPaymentTx(String),
+    ValidateHtlcError(String),
+    #[display(fmt = "Payment tx {:?} was sent to wrong address, expected {:?}", found, expected)]
+    WrongReceiverAddress {
+        found: String,
+        expected: String,
+    },
+    #[display(fmt = "Payment tx {:?} was sent from wrong address, expected {:?}", found, expected)]
+    WrongSenderAddress {
+        found: String,
+        expected: String,
+    },
+}
+
+impl From<ethabi::Error> for ValidatePaymentError {
+    fn from(err: ethabi::Error) -> Self { Self::InvalidPaymentTxData(err.to_string()) }
+}
+
+impl From<rlp::DecoderError> for ValidatePaymentError {
+    fn from(err: rlp::DecoderError) -> Self { Self::InvalidPaymentTxData(err.to_string()) }
+}
+
+impl From<keys::Error> for ValidatePaymentError {
+    fn from(err: keys::Error) -> Self { Self::InternalError(err.to_string()) }
+}
```

Comment:
You can remove this `From` implementation and map error to `InvalidInput` where `keys::Error` is used. My thinking is in the current implemented cases `keys::Error` is due to an invalid input but that might not always be the case in the future.

Reply:
ok done